### ISSUE TITLE
Using Java.extend from JavaScript erases "extra" methods and instance variables.

### ISSUE THREAD

**Issue by Charanor** [oracle/graal#2636]
**Describe GraalVM and your environment :**
- GraalVM version or commit id if built from source: **20.1.0**
- CE or EE: **CE**
- JDK version: **JDK8**
- OS and OS Version: **Windows 10**
- Architecture: **amd64**
- The output of java -Xinternalversion:
`OpenJDK 64-Bit Server VM GraalVM CE 20.1.0 (25.252-b09-jvmci-20.1-b02) for windows-amd64 JRE (1.8.0), built on Apr 21 2020 14:28:43 by "BuildBot" with MS VC++ 10.0 (VS2010)`

**Have you verified this issue still happens when using the latest snapshot?**
Yes, using graalvm-ce-java8-20.2.0-dev.

**Describe the issue**
The result of `Java.extend` does not have any methods or instance variables not present in the super class. Since there is no way to cast objects in JavaScript we can not access these values on an extended class. I have tried to find if something like this exists but it's difficult to search for documentation since the documentation is spread out between GraalVM and Nashorn (e.g. for the `Java.*` functions).

**Code snippet or code repository that reproduces the issue**
```java
// Java code
public abstract class Sayer {
    public abstract void saySomething();
}
```

---

```javascript
// JS code
const JSayer = Java.type("Sayer");
class HelloSayer {
    saySomething() {
        this.sayHello();
    }

    sayHello() {
        print("Hello!");
    }
}

const extended = Java.extend(JSayer, new HelloSayer());
// Correctly prints "Hello!" since saySomething is a method that exists on the super class.
extended.saySomething();

// TypeError: invokeMember (sayHello) on [the instance] failed due to: Unknown identifier: sayHello
extended.sayHello();
```

**Steps to reproduce the issue**
Place code above into files, compile, and run.

**Expected behavior**
A way to access "extra" methods and variables from the result of `Java.extend` that are not available in the super class. Perhaps by some type of property containing the original JavaScript object or a function that "casts" the object to the specified type.

**Additional context**
A Java library I'm using manages "components". A component is any Java Object that extends their abstract class `Component`. This is an empty abstract class only used (from what I can tell) for sanity-checking that the component being added is actually meant to be used as a component. This means that I have to `Java.extend(Component, javascript_impl)` whenever I want to add a component from JavaScript. Let's pretend the `HelloSayer` above is one of those components.

You can retrieve a certain component from the library by supplying the class of the component (the result of `Java.extend` in this case) to a function which will then return the component. We know that the value returned by the function is a wrapper around the JavaScript class `HelloSayer`. However since all "extra data" is lost when `Java.extend` happens it is impossible to access any data inside the component since, of course, from what Java is concerned the object returned is of the class `Component` and nothing else.

I'm looking for a way to access all the data from the JavaScript implementation (`saySomething` and `sayHello` in this case). Perhaps by some type of property containing the original JavaScript object or a function that "casts" the object to the specified type?:

```javascript
const Extended = Java.extend(Component, HelloSayer);
const component = new Extended();

// Maybe this
component.__jsProps.sayHello();

// Or this
Java.cast(component, HelloSayer).sayHello();
```

**Comment by woess** [comment]
20.3.0 will allow access to the underlying JS object via a special `this` member, so the following will work:
```js
const JSayer = Java.type("Sayer");
class HelloSayer {
    saySomething() {
        this.sayHello();
    }

    sayHello() {
        print("Hello!");
    }
}

const Extended = Java.extend(JSayer);
const extended = new Extended(new HelloSayer());
extended.saySomething();
extended.this.sayHello(); // extended.this instanceof HelloSayer
```
Is this satisfactory?

Note that this currently only works with:
```js
const Extended = Java.extend(JSayer);
const extended = new Extended(new HelloSayer());
```
but not class overrides:
```js
const Extended = Java.extend(JSayer, new HelloSayer());
const extended = new Extended();
```

**Comment by Charanor** [comment]
Yes that is good enough for me at least! I'm guessing it doesn't work when doing `Java.extend(JSayer, HelloSayer)` either? Thanks a lot this will definitely improve useability.

**Comment by woess** [comment]
> I'm guessing it doesn't work when doing `Java.extend(JSayer, HelloSayer)` either?

Correct. But I don't think you should not use that anyway. Creating new instances of the returned class does not create new instances of `HelloSayer` as you might expect. For that you have to use `Extended = Java.extend(JSayer), new Extended(new HelloSayer())`.
